[JRC] Fix glitch with item.extra, update detection

zuphilip opened the pull request:
* Check cases where item.extra is empty, before trying to replace
   something from there.
 * Items have now sometimes the ambiguous item type "Articles in
   periodicals and books", which is now handled as good as possible.

zuphilip commented:
@aurimasv The test cases show me a difference in the ISBNs, e.g.
```
         -   "ISBN": "978-1-4673-2754-1"
         +   "ISBN": "9781467327541"
```
However, in the real translation the slashes appear.

aurimasv commented on JRC Publications Repository.js:
```diff
@@ -67,10 +67,21 @@ function getSearchResults(doc, checkOnly) {
-		} else {//generic fallback
-			Z.debug('Unrecognized item type: ' + type[0].value);
+		if (mappingTable[type[0].textContent]) {
+			return mappingTable[type[0].textContent];
```
We should use `.value` for attributes https://developer.mozilla.org/en-US/docs/Web/API/Attr

zuphilip replied:
Okay, let me just understand this a little more: Whenever I have an xpath which ends with `@attribute`, then I should use `.value`. And for nodes (different form of xpath) we use `.textContent`. Is that correct?

aurimasv replied:
Yes,that's what we _should_ do, though we don't really follow that rule strictly. Going forward, I think we should pay more attention to this, since it may start breaking in the future.

aurimasv replied:
Also, you could do what some of the code does below and retrieve the node with the XPath, then grab the attribute via object property (`.content`) or `.getAttribute` if the property is not available. I think this would be easier to keep track of and it sometimes performs more appropriately (e.g. `.href`)

aurimasv commented:
Could you add test cases

aurimasv commented:
>@aurimasv The test cases show me a difference in the ISBNs, e.g.
```
         -   "ISBN": "978-1-4673-2754-1"
         +   "ISBN": "9781467327541"
```
However, in the real translation the slashes appear.

Yes, that's expected. The dashes are inserted right before writing to the DB (so that editing the field in Zotero client hyphenates it properly). We can't tap into that from Scaffold.

zuphilip commented:
Updated according to your comment by using the object property `.content` and the tests (ISBNs) are also updated. (I guess that some month ago, the ISBN handling was different, but good to know that everything is as expected.) Please have a look and let me know if I should squash.

aurimasv commented:
Looks ok. So were the current test cases failing before this patch? Do the current cases cover the different `detectWeb` cases? Do they cover the `item.extra` glitch?

zuphilip commented:
<blockquote> So were the current test cases failing before this patch?</blockquote>
Yes, all of them, see e.g.
```
TranslatorTester: Running 7 tests for JRC Publications Repository TranslatorTester: Running JRC Publications Repository Test 1 TranslatorTester: Translating http://publications.jrc.ec.europa.eu/repository/handle/JRC81894 Unrecognized item type: Articles in periodicals and books Unrecognized item type: Articles in periodicals and books Unrecognized item type: Articles in periodicals and books Translation using JRC Publications Repository failed: string => TypeError: item.extra is undefined stack => scrape/<@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\JRC Publications Repository.js:207:4 @chrome://zotero/content/xpcom/translation/translate_firefox.js:433:66 Zotero.Item.prototype.complete@chrome://zotero/content/xpcom/translation/translate_firefox.js:613:296 completeItem@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:205:2 init/<@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:313:4 @chrome://zotero/content/xpcom/translation/translate_firefox.js:433:66 Zotero.Item.prototype.complete@chrome://zotero/content/xpcom/translation/translate_firefox.js:613:415 doImport@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\RDF.js:1182:4 importRDF@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:358:2 doWeb/<@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:352:39 init/<@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:336:4 @chrome://zotero/content/xpcom/translation/translate_firefox.js:433:66 init@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:316:3 doWeb@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\Embedded Metadata.js:352:11 scrape/<@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\JRC Publications Repository.js:247:3 @chrome://zotero/content/xpcom/translation/translate_firefox.js:433:66 scrape@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\JRC Publications Repository.js:236:2 doWeb@C:\Users\zumstein\AppData\Roaming\Mozilla\Firefox\Profiles\fo682pah.default\zotero\translators\JRC Publications Repository.js:111:3
```
<blockquote>Do the current cases cover the different detectWeb cases?</blockquote>
The test cases cover `journalArticle` and `conferencePaper`. I don't have an example for `bookSection` but it looks naturally fitting there ;-) I will look if I can find easily an example.
<blockquote>Do they cover the item.extra glitch?</blockquote>
Hm.. I don't know it the deletion of wrongly saved tags in `item.extra` is still needed. But it seems for me just some month ago, when we added the line... Do you suggest to delete that line completely?

aurimasv commented:
OK. Looks good. Squash please.

adam3smith commented:
if you could wait 5mins before squashing. Would like to ready through some of this.

zuphilip commented:
Okay, I stopped `git rebase -i HEAD~3` just in time...

adam3smith commented:
thanks. Good to go now

zuphilip commented:
Okay, I continue the squashing of the three commits...
![squash](https://cloud.githubusercontent.com/assets/5199995/10269174/061a1086-6acf-11e5-862d-2eb3b59296e7.jpg)

aurimasv commented:
(if you're working on Windows, I would recommend TortoiseGit)

zuphilip commented:
Done.

(I used TortoiseGit before but decided to do as much as possible with the bash, which works also on all other OS or remote server.)

aurimasv commented:
>I used TortoiseGit before but decided to do as much as possible with the bash, which works also on all other OS or remote server.

True, it's good to know how to do this in the command-line, but I find TortoiseGit a lot more efficient when rebasing, blaming, looking through commit history, committing partial changes to a file, basically everything. The only thing I find myself opening up the command line for is to set the executable bit once in a while. Anyway, as long as you're aware that it exists.
